# IsoAligner

Summary: Aligning protein sequences of isoforms is necessary to transfer annotated mutational data from one splice variant to another. However, most alignment tools are fitted for homologous sequences, while simple mapping of amino acids across isoforms comes short. Here, we present an interactive alignment webservice which is tailored for the positional mapping of exons. The tool uses a customized Needleman-Wunsch algorithm including an open gap penalty combined with a gene-specific minimal exon length function and dynamically adjustable parameters. As an input, either gene/transcript/protein IDs of sequence databases (Ensembl, UniProt, RefSeq) or raw amino acid sequences, can be provided. The output of IsoAligner consists of pairwise alignments and a table of mapped amino acid positions between the canonical or supplied isoform IDs and all alternative isoforms. IsoAlignerâ€™s human isoform library comprises of over 1.3 million IDs mapped on over 120,000 protein sequences. IsoAligner, is a fast and interactive alignment tool to retrieve amino acids positions between different protein isoforms.
Availability: This tool is available as a Webservice on www.isoaligner.org. A REST API is available for programmatic access.